lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs

```diff
@@ -845,16 +873,84 @@ fromAllegraTx tx =
     -- FIXME (ADP-525): It is fine for now since we do not look at script
     -- pre-images. But this is precisely what we want as part of the
     -- multisig/script balance reporting.
-    toSLMetadata (MA.Metadata blob _scripts) = SL.MetaData blob
+    toSLMetadata (MA.AuxiliaryData blob _scripts) = SL.Metadata blob
+
+fromMaryTx
+    :: SL.Tx (Cardano.ShelleyLedgerEra MaryEra)
+    -> ( W.Tx
+       , [W.DelegationCertificate]
+       , [W.PoolCertificate]
+       )
+fromMaryTx tx =
+    ( W.Tx
+        (fromShelleyTxId $ SL.txid bod)
+        (Just $ fromShelleyCoin fee)
+        (map ((,W.Coin 0) . fromShelleyTxIn) (toList ins))
+        (map fromMaryTxOut (toList outs))
+        (fromShelleyWdrl wdrls)
+        (fromShelleyMD . toSLMetadata <$> SL.strictMaybeToMaybe mad)
+    , mapMaybe fromShelleyDelegationCert (toList certs)
+    , mapMaybe fromShelleyRegistrationCert (toList certs)
+    )
+  where
+    SL.Tx bod _wits mad = tx
+    MA.TxBody ins outs certs wdrls fee _valid _upd _adh _value = bod
+
+    -- FIXME (ADP-525): It is fine for now since we do not look at script
+    -- pre-images. But this is precisely what we want as part of the
+    -- multisig/script balance reporting.
+    toSLMetadata (MA.AuxiliaryData blob _scripts) = SL.Metadata blob
+
+    fromMaryTxOut
+         :: SL.TxOut (Cardano.ShelleyLedgerEra MaryEra)
+         -> W.TxOut
+    fromMaryTxOut (SL.TxOut addr value) =
+        W.TxOut (fromShelleyAddress addr) $
+        fromCardanoValue $ Cardano.fromMaryValue value
+
+fromCardanoValue :: Cardano.Value -> TokenBundle.TokenBundle
+fromCardanoValue = uncurry TokenBundle.fromFlatList . extract
+  where
+    extract value =
+        ( mkCoin $ Cardano.selectLovelace value
+        , convert $ Cardano.valueToList value )
+    convert = strip . map (bimap fromCardanoAssetId mkQuantity)
+
+    -- Remove ada from value.
```

Comment:
?

Reply:
@rvl ?

Reply:
I clarified the function. When constructing the TokenBundle from an asset map, it needs separate ada and non-ada assets.